### ISSUE TITLE
ZFP Compression

### ISSUE THREAD

**Issue by rabernat** [zarr-developers/numcodecs#117]
I just learned about a new compression library called ZFP: https://github.com/LLNL/zfp

> zfp is an open source C/C++ library for compressed numerical arrays that support high throughput read and write random access. zfp also supports streaming compression of integer and floating-point data, e.g., for applications that read and write large data sets to and from disk.
>
> zfp was developed at Lawrence Livermore National Laboratory and is loosely based on the algorithm described in the following paper:
>
>     Peter Lindstrom
>     "Fixed-Rate Compressed Floating-Point Arrays"
>     IEEE Transactions on Visualization and Computer Graphics
>     20(12):2674-2683, December 2014
>     doi:10.1109/TVCG.2014.2346458
>
> zfp was originally designed for floating-point arrays only, but has been extended to also support integer data, and could for instance be used to compress images and quantized volumetric data. To achieve high compression ratios, zfp uses lossy but optionally error-bounded compression. Although bit-for-bit lossless compression of floating-point data is not always possible, zfp is usually accurate to within machine epsilon in near-lossless mode.
>
> zfp works best for 2D and 3D arrays that exhibit spatial correlation, such as continuous fields from physics simulations, images, regularly sampled terrain surfaces, etc. Although zfp also provides a 1D array class that can be used for 1D signals such as audio, or even unstructured floating-point streams, the compression scheme has not been well optimized for this use case, and rate and quality may not be competitive with floating-point compressors designed specifically for 1D streams.
>
> zfp is freely available as open source under a BSD license, as outlined in the file 'LICENSE'. For more information on zfp and comparisons with other compressors, please see the zfp website. For questions, comments, requests, and bug reports, please contact Peter Lindstrom.

It would be excellent to add ZFP compression to Zarr! What would be the best path towards this? Could it be added to numcodecs?

**Comment by alimanfoo** [comment]
Just to elaborate a little, what we would like to do is implement a codec class for ZFP. The Codec interface is defined [here](https://numcodecs.readthedocs.io/en/latest/abc.html). Once there is a Python package binding ZFP, then the codec implementation is very simple, basically the ``Codec.encode(buf)`` method implementation would just pass through to a ``zfp.compress(buf)`` function, and similarly the ``Codec.decode(buf, out=None)`` method would ideally just pass through to a ``zfp.decompress(buf, out)`` function. 

There is a detail on the ``decode`` method in that numcodecs supports an ``out`` argument which can be used by compressors that have the ability to decompress directly into an existing buffer. This potentially means that decompressing involves zero memory copies. So if a ``zfp`` package offered this ability to decompress directly into a buffer exposed by a Python object via the buffer interface, this could add an additional optimisation. However, most compression libraries don't offer this, so it is not essential. I.e., if zfp does not offer this, then in numcodecs if an ``out`` argument is provided, we just do a memory copy into ``out``. For an example, see the [zlib codec implementation](https://github.com/zarr-developers/numcodecs/blob/master/numcodecs/zlib.py).

One other detail, ideally a zfp Python binding would be able to accept any object exposing the Python buffer interface. We currently also require codecs to be able to handle ``array.array`` which in Python 2 needs a special case because it doesn't implement the buffer interface. But we can work around that inside numcodecs. I.e., it is fine if a zfp binding just uses the buffer interface, no special case needed for ``array.array`` in Python 2. E.g., this is part of the reason why there is some special case code for Python 2 within the zlib codec implementation.

Hth.

**Comment by jakirkham** [comment]
> We currently also require codecs to be able to handle `array.array` which in Python 2 needs a special case because it doesn't implement the buffer interface.

We can smooth over this a bit. Opened PR ( https://github.com/zarr-developers/numcodecs/pull/119 ) as a proof-of-concept.

**Comment by lindstro** [comment]
@alimanfoo, zfp does indeed decompress directly into a user-allocated buffer.  Similarly, compression writes to a user-allocated buffer whose size zfp conservatively estimates for the user.  Alternatively, the user can specify how much memory to use for the compressed data, and zfp will ensure that the compressed data fits within the buffer (with quality commensurate with buffer size).

I know very little about zarr, but my understanding is that it partitions arrays into equal-shape chunks whose compressed storage is of variable length.  I'm guessing chunks have to be large enough to amortize the overhead of compression metadata (I see 1 MB or so recommended).  zfp provides similar functionality but uses very small chunks (4^d values for d-dimensional arrays) that are each (lossily) compressed to a fixed number of bits per chunk (the user specifies how many bits; for 3D arrays, 1024 bits is common).  I wonder if this capability could be exploited in zarr without having to rely on zfp's streaming compression interface.

**Comment by alimanfoo** [comment]
@lindstro thanks for the information, very helpful.

Apologies if I have not fully understood the details in your comment, but you are right to say that zarr partitions an array into equal-shape chunks, and then passes the data for each chunk to a compression codec (in fact this can be a pipeline of codecs configured by the user, but typically it is just a compressor). The result of encoding the chunk is then stored, and this will be of variable length depending on the data in each chunk. 

In the zarr architecture, ZFP would be wrapped as a codec, which means it could be used as the compressor for an array. So zarr would pass each chunk in the array to ZFP, and then store whatever ZFP gives it back. Zarr passes all of the data for a chunk to a codec in a single API call, so in general there is no need to use streaming compression, you can just do a one-shot encoding. Typically I find that chunks should be at a minimum 1 MB uncompressed size, usually I find upwards of 16 MB is better, depending somewhat on various factors like compression ratio and type of storage being used.

A compressor (or any other type of codec) is a black box as far as zarr is concerned. If a compressor like ZFP chooses to break the chunk down further into smaller pieces, that is an internal implementation detail. E.g., the Blosc compressor does something similar, it breaks down whatever it is given into smaller blocks, so it can then use multiple threads and compress blocks in parallel. 

If it is possible to vary the size of chunks that ZFP is using internally, then this is an option you'd probably want to expose to the user when they instantiate the ZFP codec, so they could tune ZFP for a particular dataset. 

Hth.

**Comment by lindstro** [comment]
@alimanfoo, let me try to clarify.

Whereas zarr prefers uncompressed chunks as large as 16 MB, zfp in its fixed-rate mode uses compressed chunks on the order of 8-128 bytes (a cache line or two), which provides for far finer granularity of access.  Think of zfp as a compressed floating-point format for SSE vectors and similar.  I was just thinking out loud whether a capability like that could be exploited by zarr, for example, if traversing a 3D array a 2D or 1D slice at a time, when only a very small subset of a 16 MB chunk is needed.

My reference to zfp streaming compression was meant in the sense of sequentially (de)compressing an entire array (or large portions thereof) in contrast to zfp's inline (de)compression capability, where tiny blocks (4^d scalars in d dimensions) are (de)compressed on demand in response to random access reads or writes to individual array elements.

Of course, one could use zfp as a black box codec with zarr to (de)compress large chunks, but my question had to do with whether zarr could benefit from fine-grained access to individual scalars or a few handfuls of scalars, as provided by zfp's own compressed arrays.

If not, then the best approach to adding zfp support to zarr would most likely be to use the Python bindings we're currently developing to zfp's high-level C interface, which is designed for infrequent (de)compression of the megabyte-sized chunks that zarr prefers.

**Comment by alimanfoo** [comment]
Many thanks @lindstro, very nice clarification.

> Of course, one could use zfp as a black box codec with zarr to (de)compress large chunks, but my question had to do with whether zarr could benefit from fine-grained access to individual scalars or a few handfuls of scalars, as provided by zfp's own compressed arrays.

We had a similar discussion a while back, as the blosc compressor offers something analogous, which is the ability to decompress specific items within a compressed buffer, via a function called ``blosc_getitem``. The discussion is here: https://github.com/zarr-developers/zarr/issues/40, see in particular comments from here: https://github.com/zarr-developers/zarr/issues/40#issuecomment-236844228.

The bottom line is that I think it could, in principle, be possible to modify numcodecs and zarr to leverage this kind of feature. However, it would require some reworking of the API layer between zarr and numcodecs and some non-trivial implementation work. I don't have bandwidth to do work in that direction at the moment, but if someone had the time and motivation then they'd be welcome AFAIC. 

As I understand it, the type of use case that would make this feature valuable are where you are doing a lot of random access operations into small regions of an array, i.e., pulling out individual values or small clusters of nearby values. I don't have any use cases like that personally, but maybe others do, and if so I'd be interested to know.

All my use cases involve running parallel computations over large regions of an array. For those use cases, it is fine to decompress entire chunks, as it is generally only a few chunks around the edge of an array or region where you don't need to decompress everything, and the overhead of doing a little extra decompression than strictly needed is negligible. Occasionally I do need to pull out single values or a small sub-region, but that is generally a one-off, so again the extra overhead of decompressing entire chunks is easy to live with.

> If not, then the best approach to adding zfp support to zarr would most likely be to use the Python bindings we're currently developing to zfp's high-level C interface, which is designed for infrequent (de)compression of the megabyte-sized chunks that zarr prefers.

This sounds like a good place to start.

**Comment by lindstro** [comment]
Regarding use cases, you'll have to excuse my ignorance of how zarr works and the use cases it was designed for, but here are a few that motivated the design of zfp and its small blocks:

- zfp serves as a substitute for conventional multidimensional random-accessible arrays.  zfp generally avoids the need to rewrite existing application code to traverse the arrays in an order best suited for the underlying data structure (e.g., one chunk at a time).  Ideally, you substitute only C/C++ array or STL vector declarations with zfp array declarations while leaving the remaining code intact.  Sometimes the traversal order can be hidden behind iterators, but it is common in C and even C++ to use explicit indexing and nested for loops for array computations, especially in stencil-based computations that require localized random access.  Because zfp's blocks are so small, it does not matter much in what order the (multidimensional) array is accessed, whereas if you use large chunks on the order of 16 MB, you'll want to process all elements in a chunk before moving on to the next one.  That is, if the traversal enters and exits a chunk many times, then you don't want to (de)compress the chunk each time, and you may not be able to afford caching many large uncompressed chunks.  (Does zarr cache decompressed chunks?)

- If chunks are much larger than the hardware cache size, then performance will suffer when making repeated accesses to a chunk, e.g., via kernel fusion, stencil operations, gathers and scatters between staggered grids, etc.  The compressed chunk size in zfp is usually on the order of one L1 cache line, while a decompressed chunk is a few cache lines, allowing both compressed and uncompressed data to fit in L1 cache if the access pattern exhibits good locality.  If chunks are as large as 16 MB, then the decompressed data has already been evicted from L1 and L2 cache once computation on the decompressed data is executed.

- Some applications traverse subsets of arrays.  Examples include 1D and 2D slices of a 3D array, data-dependent isocontours and integral paths, boundary layers for ghost data exchange in distributed settings, regions of interest and range queries (e.g., in visualization), subsampling (e.g., for data decimation), etc.  If chunks are too coarse, then there's a lot of overhead associated with decompressing data that is not accessed.

- zfp supports both read and write access.  Only (uncompressed and cached) blocks that are modified need to be written back to compressed storage when evicted from zfp's software cache.  If chunks are large, then a change of a single value in a chunk would trigger a lot of data to be compressed.

I do agree, however, that it is often possible to structure the traversal and computation in a streaming manner to amortize the latency associated with (de)compression.  This, however, typically requires refactoring existing code.

**Comment by alimanfoo** [comment]
Thanks @lindstro, very interesting. 

FWIW Zarr was designed primarily as a storage partner to the Dask array module, which implements a subset of the Numpy interface but as chunked parallel computations. It's also designed to work well either on a single machine (either multi-threaded or multi-process) or on distributed clusters (e.g., Pangeo-like clusters reading from object storage). With Dask, the sweet spot tends to be towards larger chunks anyway, because there is some scheduling overhead associated with each task. So there has not (yet) been any pressure to optimise access to smaller array regions. 

But as I mentioned above it is conceivable that Zarr could be modified to leverage capabilities of ZFP or Blosc to decompress chunk sub-regions. So if someone has a compelling use case and wants to explore this then please feel free.

**Comment by rabernat** [comment]
Very interesting discussion. Thanks @lindstro for taking the time to explain ZFP in such detail!

I think it is important to distinguish between what ZFP calls "chunks" and what zarr calls "chunks". ZFP chunks are clearly intended to be small. Zarr chunks tend to be much larger. An important point is that zarr is focused on serialization: **zarr chunks correspond to individual files**. It makes no sense to store millions of tiny files, due to the overhead of opening a file. This overhead becomes even more severe when the "files" are actually objects in cloud storage (a primary use case for zarr). So we would never want a 1:1 correspondence between ZFP chunks and zarr chunks.

Instead, what we are talking about here is using ZFP as a compression layer for zarr chunks. There will probably be a huge number of ZFP chunks in one zarr chunk (i.e. file). All that really matters here is:
- The compression ratio for the whole (~10 - 100 MB) zarr chunk
- The speed of compression / decompression

Although zarr in its current form will clearly not be able to leverage all of the cool features of ZFP like random access to individual elements, the bottom line is that ZFP provides high compression ratios for multidimensional floating point array data. This alone is justification for exposing it via numcodecs.

At this point, it would be great to move beyond speculation and see some actual implementation! 😉

**Comment by rabernat** [comment]
Regarding licensing, ZFP has already been released open source with a BSD-style license:
https://github.com/LLNL/zfp/blob/master/LICENSE

Not sure exactly what you need to iron out, but the terms of this license are pretty self evident. If you're vendoring ZFP within the python library, you just need to include the original license. (This is what the fpzip python library does: https://github.com/seung-lab/fpzip/tree/master/fpzip-1.2.0) If you're not vendoring ZFP but just linking against it, you don't have to do anything special in terms of license.

**Comment by kmpaul** [comment]
Perfect!  I'll have to check with @halehawk to find out what we need to do specifically.

**Comment by lindstro** [comment]
@alimanfoo, thanks for clarifying.  If zarr primarily targets I/O and secondary storage, then I think it would make sense to use one of zfp's variable-rate modes to compress the large chunks that zarr prefers.

Skimming the zarr tutorial, it seems that accessing the array is typically done via slicing, i.e., where an uncompressed copy of a subset of the array is made for subsequent, repeated access.  Is that fair?  zfp also supports slicing but creates compressed copies.  I imagine some optimizations could be made if one accesses a 2D slice of a large 3D zarr chunk compressed by zfp, since then the whole chunk would not have to be decompressed; only those zfp blocks touched by the slice require decompression.

**Comment by lindstro** [comment]
@rabernat, I agree that chunks serve an entirely different purpose in zfp and in zarr.  As I mentioned to @alimanfoo, it might make sense to exploit the fact that each zarr chunk is many independent zfp blocks, and to decompress only those zfp blocks needed when accessing a subset of the chunk.  But I imagine supporting this would likely require major surgery on zarr.

**Comment by rabernat** [comment]
> it might make sense to exploit the fact that each zarr chunk is many independent zfp blocks, and to decompress only those zfp blocks needed when accessing a subset of the chunk.

Personally I would love to see this capability in zarr! And probably so would the netCDF developers who are currently developing a zarr backend for netCDF. It would be particularly valuable for the common use case where the zarr array is chunked in the time dimension and the user wants to compute a timeseries at one single point in space. Today this basically requires reading / decompressing the entire array. Highly inefficient.

Once we get some basic zarr / ZFP integration going, it could set the stage for such future developments. Right now, my goal is to reduce our cloud storage bill as much as possible without losing scientific value.

**Comment by lindstro** [comment]
@kmpaul, I forgot that Haiying has already done some work on wrapping zfp, and I appreciate your offer to help out with this.  I will say, however, that we have a milestone due in the next few weeks to provide Python wrappers in zfp that will become part of the next official release slated for March 2019.  I have a developer currently working on this.  We'll likely make these wrappers available on a separate branch before the release if people want to play with them.

As far as integration with zarr, I think either approach would be fine.  Using the official zfPy bindings has the benefit of long-term support and addition of new zfp features as they become available, while Haiying's wrappers have already been written and could likely be used to experiment with zfp compression within zarr today.

**Comment by kmpaul** [comment]
@lindstro Yeah, that's the problem with private repos.  :smile:  I'm fine going down both paths in parallel until zfpy is available.  It should aid with more immediate testing.

**Comment by rabernat** [comment]
If anyone has any existing ZFP python code they want to try now, you can just follow the example I made with fpzip:
https://gist.github.com/rabernat/5199d80af701b442f4bec6092c8e4a70

Creating the basic codec is pretty trivial:
```python
import fpzip
from numcodecs.abc import Codec
from numcodecs import register_codec, Zstd

class FPZip(Codec):
    
    codec_id = 'fpzip'

    def __init__(self, precision=0):
        self.precision = precision

    def encode(self, buf):
        return fpzip.compress(buf, precision=self.precision)

    def decode(self, buf):
        return fpzip.decompress(buf)

    def __repr__(self):
        r = '%s(precision=%r)' % \
            (type(self).__name__,
             self.precision)
        return r
```

**Comment by jakirkham** [comment]
@lindstro, one important point is Zarr supports on "disk" storage (though this could be an object store or actually files in directories) or in-memory storage. It's possible smaller chunk size could make sense for some in-memory Zarr Array use cases. The in-memory format is a dictionary with each chunk in different key-value pairs. Fortunately whatever work goes into adding a codec for Numcodecs could be leveraged easily by either use case.

**Comment by alimanfoo** [comment]
> @rabernat, I agree that chunks serve an entirely different purpose in zfp and in zarr. As I mentioned to @alimanfoo, it might make sense to exploit the fact that each zarr chunk is many independent zfp blocks, and to decompress only those zfp blocks needed when accessing a subset of the chunk. But I imagine supporting this would likely require major surgery on zarr.

Let's just say it would be an interesting challenge :smile:. If anyone does decide they want to explore this at some point, I'd recommend having a specific, concrete use case in mind to target and benchmark, including specific computations, specific data, and a specific target platform, including what type of storage, what type of computational nodes, and what (if any) type of parallelism. I'd also recommend making sure it's an important use case, i.e., optimisation would save somebody a significant amount of time and/or money; and doing enough thinking ahead of time to be fairly convinced that the desired performance improvement should in theory be achievable. These suggestions are obvious to everyone I'm sure, but just wouldn't want anyone to sink loads of time for little reward.

**Comment by jakirkham** [comment]
Normally those sorts of things go in the constructor and are set as members of the instance. Take a look at`Zlib` for a simple example.

**Comment by rabernat** [comment]
@halehawk - a great way to move forward with this would be to just submit a pull request with your code as is. Then we can discuss the details and iterate before merging.

**Comment by halehawk** [comment]
I did a pull request yesterday.

**Comment by rabernat** [comment]
Ah fantastic!

Perhaps you could edit your PR description to include the text "Closes #117" in the description? That way it will be linked to this discussion.

**Comment by halehawk** [comment]
Do you mean I shall submit a new pull request? By the way what is ‘tox’ in the to-do list

**Comment by lindstro** [comment]
[zfp 0.5.5](https://github.com/LLNL/zfp/releases/tag/0.5.5) has been released, which adds Python bindings to the C API and the ability to (de)compress NumPy arrays.  The current release supports compression of arbitrarily strided arrays (including non-contiguous arrays), but does not yet support decompression with arbitrary strides.  This release also adds lossless compression to zfp.

We are planning on expanding zfp's Python API over the next year or so, for instance, to include Python bindings to zfp's [compressed arrays](https://zfp.readthedocs.io/en/release0.5.5/arrays.html).

**Comment by kmpaul** [comment]
Update: [zfpy](https://anaconda.org/conda-forge/zfpy) version 0.5.5 has now been released on conda-forge for Linux and OSX.

**Comment by lindstro** [comment]
Another update: [zfpy](https://pypi.org/project/zfpy/) version 0.5.5 is now available as a `pip` package for Linux, macOS, and Windows.

**Comment by kmpaul** [comment]
I believe the conda-forge version now also supports Windows.

**Comment by tasansal** [comment]
It looks like ZFP is tightly integrated with Blosc now:

https://www.blosc.org/posts/support-lossy-zfp/

It also supports Caterva style sharding out of the box.

I am gonna try this, but not sure if a version with this Blosc is in Numcodecs.

**Comment by tasansal** [comment]
> It looks like ZFP is tightly integrated with Blosc now:
> 
> https://www.blosc.org/posts/support-lossy-zfp/
> 
> It also supports Caterva style sharding out of the box.
> 
> I am gonna try this, but not sure if a version with this Blosc is in Numcodecs.

Just tried it and it is not registered in the Numcodecs version. What can we do to make that available as a part of Blosc in Numcodecs?

**Comment by halehawk** [comment]
ZFP is one of numcodecs compressor already. What are you looking for?

**Comment by tasansal** [comment]
I am mainly concerned about the Caterva functionality of Blosc. Regardless of the codec. Any idea about that being used?

Regarding ZFP being a codec already, would it not make sense to rely on Blosc's ZFP implementation instead of maintaining it in Numcodecs now that it can be considered upstream?

**Comment by lindstro** [comment]
To my knowledge, Blosc does not support zfp's reversible (lossless) mode, whereas numcodecs does.  I'd also expect that numcodecs is used by applications other than Zarr that would benefit from zfp support.

**Comment by tasansal** [comment]
> To my knowledge, Blosc does not support zfp's reversible (lossless) mode, whereas numcodecs does. I'd also expect that numcodecs is used by applications other than Zarr that would benefit from zfp support.

True, they don't support lossless.

**Comment by tasansal** [comment]
Maybe the right question I should ask is, would it be possible to support [caterva](https://github.com/Blosc/caterva) within numcodecs?

**Comment by rabernat** [comment]
We had a discussion about that in https://github.com/zarr-developers/zarr-python/issues/713. Caterva is a lot more than just a codec, and currently it does not integrate with Zarr. Also discussed at length in https://github.com/zarr-developers/zarr-python/issues/877. The basic conclusion I reach is that Caterva sits at the same level of abstraction as Zarr. It's a container, does metadata, wants to manage its own I/O. So you would either use Caterva OR Zarr, but not both.

A more suitable integration might be to implement Blosc2 in numcodecs, which would bring many of the benefits of Caterva.

**Comment by lindstro** [comment]
Let me also add that zfp itself provides similar functionality, i.e., it exposes C and C++ multidimensional array containers that support read and write random access, (de)serialization, iterators and proxy pointers, views, etc.  zfp 1.0.0 allows integrating codecs other than zfp via templates.  I have no experience with Zarr or Caterva, but my understanding is that they have a some overlap with zfp's compressed-array classes.

**Comment by tasansal** [comment]
Excellent discussion, guys, thank you for the insights.

I think this in v3 Zarr spec will also support Caterva like functionality. https://github.com/zarr-developers/zarr-python/pull/1111

spec: https://zarr-specs.readthedocs.io/en/latest/extensions/storage-transformers/sharding/v1.0.html

ZFP inside the sharded storage transformer will be very useful for our use cases!

**Comment by rabernat** [comment]
@lindstro I think your characterization is accurate. Zarr takes a modular approach at the python level (separate store object, codec, and array-interface [i.e. zarr-python]). This modularity has served us well (very easy to add a new codec or store implementation) but it prevents us from taking advantages of some of the full-stack optimizations afforded by zfp, blosc2, or caterva because they transcend multiple layers of this stack. It would be interesting to think about how to solve this... 🤔
